File: 프로그래머스/lv2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.py
def dfs(arr, x, y, l):
    start = arr[x][y]
    global zero
    global one
    
    for i in range(x, x+l):
        for j in range(y, y+l):
            if arr[i][j] != start:
                half_l = l//2
                dfs(arr, x, y, half_l)
                dfs(arr, x+half_l, y, half_l)
                dfs(arr, x, y+half_l, half_l)
                dfs(arr, x+half_l, y+half_l, half_l)
                return
            
    if start == 0:
        zero += 1
    else:
        one += 1
            
    return

def solution(arr):
    global zero
    global one
    zero = 0
    one = 0
    
    dfs(arr, 0,0, len(arr))
    
    return [zero, one]

File: 프로그래머스/2/68936. 쿼드압축 후 개수 세기/쿼드압축 후 개수 세기.java
class Solution {

    static int one = 0;
    static int zero = 0;

    public int[] solution(int[][] arr) {
        int len = arr.length;

        dfs(arr, 0, 0, len);

        return new int[]{zero, one};
    }

    public void dfs(int[][] arr, int x, int y, int len) {
        int start = arr[x][y];

        for (int i = x; i < x + len; i++) {
            for (int j = y; j < y + len; j++) {
                if (start != arr[i][j]) {
                    dfs(arr, x, y, len / 2);
                    dfs(arr, x + len / 2, y, len / 2);
                    dfs(arr, x, y + len / 2, len / 2);
                    dfs(arr, x + len / 2, y + len / 2, len / 2);
                    return;
                }
            }
        }

        if (start == 0) {
            zero += 1;
        } else {
            one += 1;
        }
    }
}